### ISSUE TITLE
Watch for edited resources in editor

### ISSUE THREAD

**Issue by supagu** [godotengine/godot#18076]
So I am writing a C++ plugin for godot. The user assigns a material to my custom Spatial node. Now I want to be notified when this material is changed in the editor.

It seems the add_change_receptor API nor the CoreStringNames::get_singleton()->changed signal are emitted when say the materials albedo colour changes for example.
So my next avenue was to try to connect to the property editor "variant_changed" signal, but there seems to be no way to determine what object the Property editor is editing.

I would like some way to watch when the material is modified.

**Comment by Noshyaar** [comment]
For change_receptors to work, the Resource must call `_change_notify(string)` when it's changed. See TileSet for example.

https://github.com/godotengine/godot/blob/0ee72fbd30b25ff397b570294c3cd2f65c176e34/scene/resources/tile_set.cpp#L310

**Comment by supagu** [comment]
yes, the material does not call this which i would expect when the colour is changed.

**Comment by akien-mga** [comment]
For the reference, the idea for 4.0 is to drop the change receptor API which is pretty hacky, and instead make the Inspector poll for changes constantly (which could then be notified to the EditorInterface/EditorInspector API I guess for plugin use).

```
14:36 <reduz> Akien: the change receptor API needs to go away
14:36 <reduz> and the inspector needs to constantly poll for changes
14:37 <reduz> this will simplify things enormously, the problem is that when I tried to do it, I found we merged plenty of PRs of contributors hacking the change receptors for other things not the inspector that will need to be rewritten in a different way and it was quite some work
14:37 <reduz> so I never got around to fix this
14:37 <reduz> it has been a problem for a long time now
```

**Comment by Calinou** [comment]
`_change_notify()` was removed in `master`, but I don't know if an API for plugins is exposed yet.

**Comment by Zylann** [comment]
@Calinou I'm not sure it relates to this issue. `property_list_changed` and `_change_notify` was for notifying the inspector that the object's properties changed.

But this issue is about the other side, where you have a reference to a resource, and want to know when it changed. The `changed` signal exists, but apparently is not always emitted. Alternatives exist but weren't working either in these cases.

> It seems the add_change_receptor API nor the CoreStringNames::get_singleton()->changed signal are emitted when say the materials albedo colour changes for example.

> I would like materials to give me this information too when they change, especially ShaderMaterials with custom properties

**Comment by KoBeWi** [comment]
> The changed signal exists, but apparently is not always emitted.

I think the signal should be emitted on all Resource changes. If it doesn't in some cases, it's a bug IMO.

**Comment by outfrost** [comment]
I just ran into this issue when building a workaround tool for Godot not having resource inheritance.

I'm creating material variations by duplicating a material resource and changing some of its properties.

I need my tool to know when properties of the original material change, so that I can either replicate the changes on the copy, or just make a new duplicate and re-apply my overrides.

It appears this is currently impossible.

I'm using version 3.6 for this project, but from what I can tell, the situation hasn't meaningfully changed in 4.x, besides documentation being edited to a very unhelpful "usually".
